### PR TITLE
Add better type guards

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
Geez rustc and trybuild are really falling over

File: hydroflow_datalog/src/snapshots/hydroflow_datalog__tests__join_with_other@datalog_program.snap [hunk 1]
```diff
@@ -6,67 +6,29 @@ fn main() {
     {
         {
             use hydroflow::{var_args, var_expr};
+            #[allow(unused)]
```
Comment (jhellerstein) [inline hydroflow_datalog/src/snapshots/hydroflow_datalog__tests__join_with_other@datalog_program.snap:9]: What's all this?

Comment (jhellerstein) [reply hydroflow_datalog/src/snapshots/hydroflow_datalog__tests__join_with_other@datalog_program.snap:9]: Ie why is this affecting the datalog outputs?

File: hydroflow_lang/src/graph/ops/join.rs [hunk 2]
```diff
@@ -38,7 +38,25 @@ pub const JOIN: OperatorConstraints = OperatorConstraints {
         let lhs = &inputs[0];
         let rhs = &inputs[1];
         let write_iterator = quote_spanned! {op_span=>
-            let #ident = #root::compiled::pull::SymmetricHashJoin::new(#lhs, #rhs, &mut #joindata_ident);
+            let #ident = {
+                /// Limit error propegation by bounding locally, erasing output iterator type.
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/join.rs:42]: comment typo: "propagation"

File: hydroflow_lang/src/graph/ops/join.rs [hunk 3]
```diff
@@ -38,7 +38,25 @@ pub const JOIN: OperatorConstraints = OperatorConstraints {
...
+                #[inline(always)]
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/join.rs:43]: I don't pretend to understand this.

**Review by jhellerstein — APPROVED** [review]
LGTM though see comments. The new error outputs are way nicer!

**Comment by MingweiSamuel** [comment]
and rustfmt too ...

**Pull request merged**